Cancel taskrun using entrypoint binary

adshmh opened the pull request:
<!-- 🎉🎉🎉 Thank you for the PR!!! 🎉🎉🎉 -->

# Changes

The cancellation of taskruns is now done through the entrypoint binary
through a new flag called 'cancel_file'. This removes the need for
deleting the pods to cancel a taskrun, allowing examination of the logs
on the pods from cancelled taskruns. Part of work on issue #3238

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [ ] [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) included if any changes are user facing
- [x] [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including
  functionality, content, code)
- [x] Release notes block below has been filled in or deleted (only if no user facing changes)

# Release Notes

<!--
Describe any user facing changes here, or delete this block.

Examples of user facing changes:
- API changes
- Bug fixes
- Any changes in behavior
- Changes requiring upgrade notices or deprecation warnings

For pull requests with a release note:

For pull requests that require additional action from users switching to the new release, include the string "action required" (case insensitive) in the release note:

```release-note
action required: your release note here
```

For pull requests that don't need to be mentioned at release time, use the `/release-note-none` Prow command to add the `release-note-none` label to the PR. You can also write the string "NONE" as a release note in your PR description:

```release-note
NONE
```
-->

```release-note
Pods corresponding to a cancelled taskrun are no longer deleted: they are stopped instead.
```

pritidesai commented:
thanks @adshmh for this! Welcome to the community! 🎉 
/ok-to-test

vdemeester commented:
/assign
/hold

@adshmh thanks for this PR.  From reading it at a high level, it doesn't change the way a user ask for cancel (through update `spec.status`, but only how the taskrun reconciler handles the cancel.

I am all for that idea, but it might affect current user somehow, so I wonder if we should have this behind a feature flag that we switch on by default after a few release, wdyt ?

/cc @tektoncd/core-maintainers

adshmh commented:
> /assign /hold
> 
> @adshmh thanks for this PR. From reading it at a high level, it doesn't change the way a user ask for cancel (through update `spec.status`, but only how the taskrun reconciler handles the cancel.
> 
> I am all for that idea, but it might affect current user somehow, so I wonder if we should have this behind a feature flag that we switch on by default after a few release, wdyt ?

Thank you for pointing this out. I will add a feature flag for the new behavior of cancel.

imjasonh commented on cmd/entrypoint/waiter.go:
```diff
@@ -31,11 +32,14 @@ func (rw *realWaiter) setWaitPollingInterval(pollingInterval time.Duration) *rea
 	}
 	for ; ; time.Sleep(rw.waitPollingInterval) {
+		if ctx.Err() != nil {
+			return nil
```
```suggestion
			return err
```

Or probably ignore it `if !errors.Is(err, context.Canceled)`

adshmh replied:
Thank you for the review. Fixed.

imjasonh commented on cmd/entrypoint/waiter_test.go:
```diff
@@ -38,7 +39,7 @@ func TestRealWaiterWaitMissingFile(t *testing.T) {
 	doneCh := make(chan struct{})
 	go func() {
-		err := rw.setWaitPollingInterval(testWaitPollingInterval).Wait(tmp.Name(), false, false)
+		err := rw.setWaitPollingInterval(testWaitPollingInterval).Wait(context.Background(), tmp.Name(), false, false)
```
For readability, maybe `ctx := context.Background()` up at the top of this test, and reuse it throughout.

adshmh replied:
Thank you for the review. Fixed.

imjasonh commented on pkg/entrypoint/entrypointer.go:
```diff
@@ -114,7 +117,7 @@ func (e Entrypointer) Go() error {
 
 	for _, f := range e.WaitFiles {
-		if err := e.Waiter.Wait(f, e.WaitFileContent, e.BreakpointOnFailure); err != nil {
+		if err := e.Waiter.Wait(context.Background(), f, e.WaitFileContent, e.BreakpointOnFailure); err != nil {
```
Could we have `Go` take a `context.Context` and use that here? If it's still a `context.Background()` inside `cmd/entrypoint/main.go` that's fine, it's just a bit cleaner and makes it easier to add other stuff like cancelling when we get SIGINT, which we might want to do later.

adshmh replied:
Thank you for the review. Fixed.

imjasonh commented on pkg/entrypoint/entrypointer_test.go:
```diff
@@ -160,6 +165,18 @@ func TestEntrypointer(t *testing.T) {
 		breakpointOnFailure: true,
+	}, {
+		desc:       "Runner completes if not cancelled",
+		cancelFile: ".",
```
Can we make this `cancelfile` or something? `"."` is an odd name for this file.

adshmh replied:
Thank you for the review. Fixed.

imjasonh reviewed (commented):
Wow, what a great change! 🎉 

Thanks for your contribution, +1 to Vincent's point about a feature flag, but otherwise, this looks great.

adshmh commented:
> /assign /hold
> 
> @adshmh thanks for this PR. From reading it at a high level, it doesn't change the way a user ask for cancel (through update `spec.status`), but only how the taskrun reconciler handles the cancel.
> 
> I am all for that idea, but it might affect current user somehow, so I wonder if we should have this behind a feature flag that we switch on by default after a few release, wdyt ?
> 
> /cc @tektoncd/core-maintainers

Thank you for the review. The new feature of cancelling using the entrypoint binary is now behind a feature flag, with default set to false.

vdemeester commented on cmd/entrypoint/waiter.go:
```diff
@@ -31,11 +32,14 @@ func (rw *realWaiter) setWaitPollingInterval(pollingInterval time.Duration) *rea
 		return nil
 	}
 	for ; ; time.Sleep(rw.waitPollingInterval) {
+		if err := ctx.Err(); err != nil {
```
Do we also want to handle `ctx.Done()` ? 🤔 Not sure if that case would happen but…

adshmh replied:
Thank you for the review. Fixed.
Looking at the code again, I think using your suggestion, i.e. `ctx.Done()` is more readable. Updated.

On a side note, I increased the polling interval on the unit tests from 10 to 25 milliseconds, as it seems to reduce flakes significantly (from about 5% to around 0,5% in my tests). If this needs to be reverted, please let me know.

`const testWaitPollingInterval = 25 * time.Millisecond`

vdemeester commented on pkg/entrypoint/entrypointer.go:
```diff
@@ -142,19 +145,46 @@ func (e Entrypointer) Go() error {
+		errChan := make(chan error, 1)
+		go func() {
+			errChan <- e.Runner.Run(ctx, e.Command...)
+			cancel()
```
Does it mean we'll run `cancel` twice ? (here and the `defer`)
I don't remember if it's a no-op or if it panics.. If it's a no-op we are fine though.

adshmh replied:
Thank you for the review. Yes, we will end up calling `cancel` twice on the context, but that is no-op ([I tested to be sure](https://go.dev/play/p/PKt2M6X7inR)).
I added the `defer cancel()` for both readability and ensuring clean-up. If removing it is preferred, please let me know.

vdemeester reviewed (commented):
2 small questions / comments, otherwise, looks good 👍🏼

adshmh commented:
/retest

osherdp commented:
@adshmh @vdemeester any idea what's left for this change to get in?
If it will have positive effects on #4035 then it will solve lots of problems for us :smiley:

lbernick reviewed (commented):
Thanks for this @adshmh! When a pipelineRun is cancelled, it will also cancel its child TaskRuns. I'm not sure if any of the PipelineRun cancellation logic needs to be changed, but could you add some tests to the PipelineRun reconciler for this new cancellation strategy?

In addition, could you please add some docs on the new feature flag, and update the release note to specify that this behavior is controlled by the feature flag?

vdemeester commented:
> Thanks for this @adshmh! When a pipelineRun is cancelled, it will also cancel its child TaskRuns. I'm not sure if any of the PipelineRun cancellation logic needs to be changed, but could you add some tests to the PipelineRun reconciler for this new cancellation strategy?

From the `PipelineRun` perspective, nothing changes 👼🏼

vdemeester commented on pkg/entrypoint/entrypointer.go:
```diff
@@ -142,19 +145,46 @@ func (e Entrypointer) Go() error {
 				Value:      "TimeoutExceeded",
 				ResultType: v1beta1.InternalTektonResultType,
 			})
+		} else if cancelled {
```
@lbernick I think the name is a bit infortunate as it is this struct that we use for returning `Result` I think 🙃 
We could rename it at some point.

afrittoli commented:
This is unfortunately not ready to merge, so I'll move it to the next milestone

adshmh commented:
> /assign I'll try to carry this patch later this week 👼🏼

Sorry for the delay on this. I will follow up by addressing the remaining concerns in a few days.

adshmh commented on pkg/entrypoint/entrypointer.go:
```diff
@@ -142,19 +145,46 @@ func (e Entrypointer) Go() error {
+
+		var cancelled bool
+		if e.CancelFile != "" {
+			if err := e.Waiter.Wait(ctx, e.CancelFile, true, e.BreakpointOnFailure); err != nil {
```
Thank you for the review. Indeed the logic here could benefit from some simplification. I will do another review to see if it can be simplified without making too many modifications elsewhere.

vdemeester commented:
Carrying on https://github.com/tektoncd/pipeline/pull/5401

jerop commented:
doing a clean up of stale pull requests - feel free to reopen if you pick up this work again

newer pull request in https://github.com/tektoncd/pipeline/pull/5401

/close
